sql-composer/src/parser.rs

```diff
@@ -29,45 +31,50 @@ use std::str::FromStr;
 pub fn template(
     span: Span,
     alias: Option<SqlCompositionAlias>,
-) -> IResult<Span, ParsedItem<SqlComposition>> {
+) -> Result<ParsedItem<SqlComposition>> {
```

Comment:
> big fan of commit "break down remaining macro-based parsers." Excellent to see the function forms, I find them easier to follow than the macro syntax.

Yeah, some of those parsers were getting out of control. I'll miss fold_many!() though...

> I'll need you to explain the move back to Result from IResult + Span to me at some point so I "get it". Thx

The IResult is a nom type that return an Ok/Error along with what is remaining to parse. We are using Span from [https://github.com/fflorent/nom_locate](nom_locate) which provides some additional tracking so that we can get good positional information of where things came from (positional information we put in the ParsedItem wrappers). 

For `template`, it is the entry point the composers use to start parsing and should never be called by any other parsers. So, since the composers are calling it, I needed to be able to use `crate::error::Result` so that it could bubble up to the caller of the composer along with other errors like IO errors when we can't find the file. 

Since the crate::error::Result stuff is all [https://docs.rs/error-chain/0.12.1/error_chain/](error_chain), It should also allow us to chain all the different errors together and do much nicer error handling once I understand error_chain better. The `error_chain` lib is really nice for converting between other crates but I had a hard time converting from them.

I think I like this divide in Result and IResult at the template parser, which I think is a sign that it doesn't belong in lib/parser.rs anymore. I think we want to move it into the Composer trait with a better function name.